nbconvert/preprocessors/execute.py

```diff
@@ -242,45 +290,46 @@ def preprocess(self, nb, resources):
         path = resources.get('metadata', {}).get('path', '')
         if path == '':
             path = None
-
+        self.nb = nb
         # clear display_id map
         self._display_id_map = {}
 
-        # from jupyter_client.manager import start_new_kernel
-
-        def start_new_kernel(startup_timeout=60, kernel_name='python', **kwargs):
-            km = self.kernel_manager_class(kernel_name=kernel_name)
-            km.start_kernel(**kwargs)
-            kc = km.client()
-            kc.start_channels()
-            try:
-                kc.wait_for_ready(timeout=startup_timeout)
-            except RuntimeError:
-                kc.stop_channels()
-                km.shutdown_kernel()
-                raise
-
-            return km, kc
-
-        kernel_name = nb.metadata.get('kernelspec', {}).get('name', 'python')
-        if self.kernel_name:
-            kernel_name = self.kernel_name
-        self.log.info("Executing notebook with kernel: %s" % kernel_name)
-        self.km, self.kc = start_new_kernel(
-            startup_timeout=self.startup_timeout,
-            kernel_name=kernel_name,
-            extra_arguments=self.extra_arguments,
-            cwd=path)
-        self.kc.allow_stdin = False
-        self.nb = nb
-
+        self.km, self.kc = self.start_new_kernel(cwd=path)
         try:
-            nb, resources = super(ExecutePreprocessor, self).preprocess(nb, resources)
+            yield
```

Comment:
Should we yield the context variables `nb, km, kc` for more easier understanding and variable consumption downstream?

Reply:
We can yield the context variables… but the problem is that we still need all three (notebook (`nb`), kernel client (`kc`) and kernel mangager (`km`)) to be accessible via `self` later on:

https://github.com/mpacer/nbconvert/blob/0de8f30404a99e63d5cc22e31cb4a984cc201beb/nbconvert/preprocessors/execute.py#L374-L375
```python
        for cell_idx, output_indices in self._display_id_map[display_id].items():
            cell = self.nb['cells'][cell_idx]
```

https://github.com/mpacer/nbconvert/blob/0de8f30404a99e63d5cc22e31cb4a984cc201beb/nbconvert/preprocessors/execute.py#L392
```python
                msg = self.kc.shell_channel.get_msg(timeout=timeout)
```


https://github.com/mpacer/nbconvert/blob/0de8f30404a99e63d5cc22e31cb4a984cc201beb/nbconvert/preprocessors/execute.py#L413-L427
```python
    def run_cell(self, cell, cell_index=0):
        msg_id = self.kc.execute(cell.source)
        self.log.debug("Executing cell:\n%s", cell.source)
        exec_reply = self._wait_for_reply(msg_id, cell)

        outs = cell.outputs = []

        while True:
            try:
                # We've already waited for execute_reply, so all output
                # should already be waiting. However, on slow networks, like
                # in certain CI systems, waiting < 1 second might miss messages.
                # So long as the kernel sends a status:idle message when it
                # finishes, we won't actually have to wait this long, anyway.
                msg = self.kc.iopub_channel.get_msg(timeout=self.iopub_timeout)
```

https://github.com/mpacer/nbconvert/blob/0de8f30404a99e63d5cc22e31cb4a984cc201beb/nbconvert/preprocessors/execute.py#L393-L399
```python
            except Empty:
                self.log.error(
                    "Timeout waiting for execute reply (%is)." % self.timeout)
                if self.interrupt_on_timeout:
                    self.log.error("Interrupting kernel")
                    self.km.interrupt_kernel()
                    break
```

So I don't know how helpful it will be to yield them only to then ignore them.

Reply:
I'm up for making a larger refactor to remove this dependence on instance attributes, but that should be in another PR.

In the meantime do you still want me to yield these for better downstream consumption and easier understanding even if they will immediately be assigned to `self`? (also, because of the way that the dynamic default works for `kernel_name` we need to make sure that `self.nb` is assigned before we start up the kernel (or we need to manually extract the kernel name and pass it through).